[lldb] Add Checksum to FileSpec

JDevlieghere opened the pull request:
Store a Checksum in FileSpec. Its purpose is to store the MD5 hash added
to the DWARF 5 line table.

JDevlieghere commented:
Depends on #71456. 

I considered an alternative where we don't store the `Checksum` in the `FileSpec` but create a new `SupportFile` class that wraps a `Checksum` and a `FileSpec`. The reason I didn't got with this approach is because of the `FileSpecList`. The latter has helper functions that are used in the context of "support files" which would have to duplicated in a `SupportFileList` and I wasn't sure we weren't vending a `FileSpecList` through the SB API that would now become a `SupportFileList`. 

A checksum is 16 bytes. Please let me know if you think that the overhead of that is too much and outweighs the churn and code duplication to adopt a `SupportFile` and `SupportFileList`.

bulbazord commented on lldb/include/lldb/Utility/FileSpec.h:
```diff
@@ -72,7 +75,8 @@ class FileSpec {
   ///
   /// \see FileSpec::SetFile (const char *path)
-  explicit FileSpec(llvm::StringRef path, Style style = Style::native);
+  explicit FileSpec(llvm::StringRef path, Style style = Style::native,
+                    const Checksum &checksum = {});
```
Update the doxygen

bulbazord reviewed (commented):
You pushed a commit from a different PR btw

medismailben commented:
> Depends on #71456.
> 
> I considered an alternative where we don't store the `Checksum` in the `FileSpec` but create a new `SupportFile` class that wraps a `Checksum` and a `FileSpec`. The reason I didn't got with this approach is because of the `FileSpecList`. The latter has helper functions that are used in the context of "support files" which would have to duplicated in a `SupportFileList` and I wasn't sure we weren't vending a `FileSpecList` through the SB API that would now become a `SupportFileList`.
> 
> A checksum is 16 bytes. Please let me know if you think that the overhead of that is too much and outweighs the churn and code duplication to adopt a `SupportFile` and `SupportFileList`.

I personally like the current implementation, I think it makes sense to have the `Checksum` in the `FileSpec`, I don't see the point of making new `SupportFile{,List}` classes if it's just wrapping a `FileSpec` and its `Checksum`.

medismailben commented on lldb/include/lldb/Utility/FileSpec.h:
```diff
@@ -13,15 +13,18 @@
 #include "llvm/ADT/StringRef.h"
 #include "llvm/Support/FileSystem.h"
 #include "llvm/Support/FormatVariadic.h"
+#include "llvm/Support/MD5.h"
```
nit: you shouldn't need this since it's already included in `Checksum.h`

medismailben commented on lldb/include/lldb/Utility/FileSpec.h:
```diff
@@ -420,13 +428,17 @@ class FileSpec {
   /// Called anytime m_directory or m_filename is changed to clear any cached
   /// state in this object.
   void PathWasModified() {
+    m_checksum = Checksum();
```
Would be nice to recompute the checksum if the path was changed.

JDevlieghere replied:
Given that the FileSpec is a system independent path description (i.e. the file doesn't need to exist on the host) there's no way to recompute it automatically, although the user can specify a new checksum if the path changes.

medismailben reviewed (approved):
LGTM with comments.

medismailben commented:
> So I debugged lldb with lldb and adding a counter into the FileSpec constructor and came up a max of 721833 files after trying to set a breakpoint in Driver.cpp, so that storage is about 16.5MB which isn't too bad. This goes up to 27.5MB with the FileSpec being 40 bytes, so not too bad. Hopefully this won't overwhelm some larger targets in LLDB memory wise

@clayborg I think it would be worth having that stat for FileSpecs in-tree so everyone can have access to it using `statistics dump` for instance.

clayborg commented:
> > So I debugged lldb with lldb and adding a counter into the FileSpec constructor and came up a max of 721833 files after trying to set a breakpoint in Driver.cpp, so that storage is about 16.5MB which isn't too bad. This goes up to 27.5MB with the FileSpec being 40 bytes, so not too bad. Hopefully this won't overwhelm some larger targets in LLDB memory wise
> 
> @clayborg I think it would be worth having that stat for FileSpecs in-tree so everyone can have access to it using `statistics dump` for instance.

I agree. Easy to add a static variable into FileSpec to do the counting and then hookup to stats. Might be nice to make a mix in class that any internal objects can inherit from that wouldn't change the size of any objects, but on construction would add to a count, and on destruction would decrement. It would be nice to track FileSpec, Symbol and other common objects in a debugger process.

JDevlieghere committed:
[lldb] Add Checksum to FileSpec

Store a Checksum in FileSpec. Its purpose is to store the MD5 hash that
was added to the DWARF 5 line table.

This increases the size of a FileSpec from 24 to 40 bytes. The
alternative is to introduce a new SupportFile abstraction for a FileSpec
+ Checksum but that would require a corresponding SupportFileList class.
During review we decided that wasn't worth it, but that's something we
can revisit in the future.
